shadowsocks/dialer.go

```diff
@@ -0,0 +1,165 @@
+package shadowsocks
+
+import (
+	"io"
+	"errors"
+	"net"
+	"strconv"
+	"sync"
+
+	onet "github.com/Jigsaw-Code/outline-ss-server/net"
+	"github.com/shadowsocks/go-shadowsocks2/core"
+	"github.com/shadowsocks/go-shadowsocks2/shadowaead"
+	"github.com/shadowsocks/go-shadowsocks2/socks"
+)
+
+// Dialer is a dialer for Shadowsocks proxy connections.
+type Dialer interface {
+	// DialTCP connects to `address` over TCP though the Shadowsocks proxy.
+	// `address` has the form `host:port`, where `host` can be a domain name or IP address.
+	DialTCP(address string) (onet.DuplexConn, error)
+	// DialUDP relays UDP packets to/from `address` though the Shadowsocks proxy.
+	// `address` has the form `host:port`, where `host` can be a domain name or IP address.
+	DialUDP(address string) (onet.PacketConn, error)
+}
+
+type ssDialer struct {
+	proxyIP   net.IP
+	proxyPort int
+	cipher    shadowaead.Cipher
+}
+
+// NewDialer creates a Dialer that routes connections to a Shadowsocks proxy listening at
+// `host:port`, with authentication parameters `cipher` (AEAD) and `password`.
+func NewDialer(host, password, cipher string, port int) (Dialer, error) {
+	proxyIP, err := net.ResolveIPAddr("ip", host)
```

Comment:
I like your idea, but I'm not convinced this is the appropriate layer to implement it. The current interface allows the client to create multiple dialers with every proxy address.

Reply:
`net.Dial` and `DialContext` do happy eyeballs and fallback automatically, so one option would be to just pass the host unmodified and let `Dial` handle it.  (This would require using `Dial("tcp"...)` and then asserting the result to `TCPConn`.)

Reply:
Another implication of this change is that we loose the ability to bind to a local address.

Reply:
@fortuna, added the TODO in b847f8b